File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    songs = {}
    for x in range(len(genres)):
        if genres[x] not in songs:
            songs[genres[x]] = {'song':[(x, plays[x])], 'total': plays[x]}
        else:
            songs[genres[x]]['song'] += [(x, plays[x])]
            songs[genres[x]]['total'] += plays[x]
    songs = sorted(songs.items(), key = lambda x : x[1]['total'], reverse=True)
    for x in range(len(songs)):
        songs[x][1]['song'] = sorted(songs[x][1]['song'], key = lambda x : x[1], reverse=True)
        for y in range(2):
            if len(songs[x][1]['song']) == 1:
                answer.append(songs[x][1]['song'][0][0])
                break
            answer.append(songs[x][1]['song'][y][0])
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    class Song{
        int id;
        int play;
        
        public Song(int id, int play){
            this.id = id;
            this.play = play;
        }
    }
    
    class Genre{
        List<Song> songs;
        String genre;
        int total;
        public Genre(String genre){
            songs = new ArrayList<>();
            this.genre = genre;
            this.total = 0;
        }
        
        public void add(int id, int play){
            songs.add(new Song(id, play));
            total += play;
        }
    }
    public int[] solution(String[] genres, int[] plays) {
        int[] answer = {};
        int id = 0;
        Map<String, Genre> map = new HashMap<>();
        for(int i = 0 ; i < genres.length ; i++){
            map.putIfAbsent(genres[i], new Genre(genres[i]));
            map.get(genres[i]).add(id++, plays[i]);
        }
        
        for(String key : map.keySet()){
            map.get(key).songs.sort((a, b) -> {
                if(b.play == a.play){
                    return a.id - b.id;
                }
                return b.play - a.play;
            });
        }
        
        List<String> keys = new ArrayList<>(map.keySet());    
        keys.sort((a, b) ->{
            return map.get(b).total - map.get(a).total;
        });
        List<Integer> ans = new ArrayList<>();
        for(String key : keys){
            // for(Song s : map.get(key).songs){
            //     System.out.println(s.id + " " + s.play);
            // }
            // System.out.println(map.get(key).total);
            ans.add(map.get(key).songs.get(0).id);
            if(map.get(key).songs.size() > 1){
                ans.add(map.get(key).songs.get(1).id);
            }
        }
        answer = new int[ans.size()];
        for(int i = 0 ; i < ans.size() ; i++){
            answer[i] = ans.get(i);
        }
        return answer;
    }
}